feature(backport): backport grapher variable to walden

Marigold opened the pull request:
TBD

Marigold commented:
@danyx23 (and perhaps @larsyencken as well) could you take a brief look and tell me whether this is a go or no-go, please? I think it's in a shape to align our visions. I've tried to summarize everything in the description, so no need to read through code.

danyx23 commented:
Cool! I'll have a look at this tomorrow

larsyencken commented:
From our discussion: it should be fine to export whole datasets in one go, let's try that as our first implementation.

danyx23 reviewed (commented):
I think this is looking pretty good already. The TODO's open a lot of interesting questions that are probably easier to tackle in a call.

To your big questions:
I think the dataset level is good initally. If we discover that we really need variable level in the future then we can implement that later.

I think adding a trigger on data_values is better than hashing as we pay the perf penalty once on write instead every time on read.

I don't understand the walden index writing issue - to discuss in the call.

The date in the walden path is indeed weird. Maybe for backports we should make it possible to do without? The dataset id I find pretty useful.

Yes, garden level processing seems to be a good step for breaking this up in smaller chunks.

Marigold commented on backport/backport.py:
````diff
@@ -0,0 +1,288 @@
+    An example (taken from https://stackoverflow.com/questions/4753878/how-to-program-a-mysql-trigger-to-insert-row-into-another-table)
+    ```
+    delimiter #
+    create trigger variable_changelog_trig after insert on data_values
````
> My gut feeling is that the trigger is the better way to go here

Agree, content hash on values is a very dumb way to check changes. However, I checked that we'd be importing 169M data values if we wanted all datasets with at least one variable in a chart which would take just 12 mins. I was expecting like 12 hours... Given that importing actual data values takes much longer, content hashing might not be a bottleneck after all. Worth a discussion for sure.

I've tested content hashing on 7.7M values and memory was fine (due to upgrade to MySQL 8 for sure :) ). Given that the largest dataset has 11M values, this shouldn't be a problem.

Marigold commented on backport/backport.py:
```diff
@@ -0,0 +1,288 @@
+
+    # upload config to walden
+    log.info("backport.upload_config")
+    _upload_config_to_walden(
```
I started implementing this and then realized it might be more trouble than it's worth. If I added optional `--upload` then what could happen is that:
- you run `backport` which updates your walden index with md5 of the file, but don't upload anything
- you run `backport --upload` hoping it would upload files to walden, but we find that md5 of the file is identical to md5 from walden index and it skip the dataset

When you start thinking about it, you realize that if we don't have md5 stored in S3 next to the file, there's no easy way how to do this safely.

However! We have `--dry-run` that lets you run the backport without adding it to walden index. That worked well for my development.

danyx23 replied:
Ok cool, this will do then!

Marigold commented:
@danyx23 since your last review I've fixed most of `TODO:`s I had in code, implemented things we discussed last time and also added Sources to variables and datasets. There are also two PRs in submodules https://github.com/owid/walden/pull/19 and https://github.com/owid/owid-catalog-py/pull/27.

I've created tasks for the remaining issues, so once we merge this one I can continue with smaller PRs.

danyx23 reviewed (approved):
Looks good!
